internal/profiling/profiling.go

```diff
@@ -0,0 +1,251 @@
+	Metadata []byte
+	// A collection of Timers and a mutex for append operations on the slice.
+	mu     sync.Mutex
+	Timers []Timer
+}
+
+// A power of two that's large enough to hold all timers within an average RPC
+// request (defined to be a unary request) without any reallocation. A typical
+// unary RPC creates 80-100 timers for various things. While this number is
+// purely anecdotal and may change in the future as the resolution of profiling
+// increases or decreases, it serves as a good estimate for what the initial
+// allocation size should be.
+const defaultStatAllocatedTimers int32 = 128
+
+// NewStat creates and returns a new Stat object.
+func NewStat(statTag string) *Stat {
+	return &Stat{
+		StatTag: statTag,
+		Timers:  make([]Timer, 0, defaultStatAllocatedTimers),
+	}
+}
+
+// NewTimer creates a Timer object within the given stat if stat is non-nil.
+// The value passed in timerTag will be attached to the newly created Timer.
+// NewTimer also automatically sets the Begin value of the Timer to the current
+// time.
+//
+// The user is expected to call stat.Egress with the returned index as argument
+// to mark the end. The return value is not a pointer to a Timer object because
+// the internal slice may be re-allocated freely, which would make the
+// reference to pointers of the past obsolete.
+//
+// Why a slice of timer? To make NewTimer allocation-free, we don't allocate a
+// new timer every time. Instead a slice of timers large enough to hold most
+// data is created and elements from this slice are used. This should put less
+// pressure on the GC too.
+func (stat *Stat) NewTimer(timerTag string) int {
+	if stat == nil {
+		return 0
+	}
+
+	t := time.Now()
+	stat.mu.Lock()
+	stat.Timers = append(stat.Timers, Timer{
+		TimerTag: timerTag,
+		GoID:     goid(),
+		Begin:    t,
+	})
+	index := len(stat.Timers) - 1
+	stat.mu.Unlock()
+	return index
+}
+
+// Egress marks the completion of a given timer within a stat.
+func (stat *Stat) Egress(index int) {
+	if stat == nil {
+		return
+	}
+
+	t := time.Now()
+	stat.mu.Lock()
+	if index < len(stat.Timers) {
+		stat.Timers[index].End = t
+	}
+	stat.mu.Unlock()
+}
+
+// AppendTimer appends a given Timer object to the internal slice of timers. A
+// deep copy of the timer is made (i.e. no reference is retained to this
+// pointer) and the user is expected to lose their reference to the timer to
+// allow the Timer object to be garbage collected.
+func (stat *Stat) AppendTimer(timer *Timer) int {
+	if stat == nil || timer == nil {
+		return 0
+	}
+
+	stat.mu.Lock()
+	stat.Timers = append(stat.Timers, *timer)
+	index := len(stat.Timers) - 1
+	stat.mu.Unlock()
+	return index
+}
+
+// ServerConnectionCounter counts the number of connections a server has seen.
+// This counter is embedded within a StreamStat's Metadata along with each
+// stream's stream ID to uniquely identify a query. Accessed atomically.
+var ServerConnectionCounter uint64
```

Comment:
I'm assuming this is `++` by the server, and then read by the server (or the transport). And the stat metadata is `<connectionID><streamID>`.

We should make this an internal detail of the transport I think.
Connection counter should be the same as transport counter (because transport and connection are 1:1 mapping).
For the metadata field, we can treat it as an opaque stream ID generated by transport, and is guaranteed to be unique. The transport can still generate it the same way `<transportID><streamID>`.

Reply:
Agreed, done.